net/libfab/libfab.c

```diff
@@ -1295,12 +1313,86 @@ static int libfab_buf_register(struct m0_net_buffer *nb)
  */
 static int libfab_buf_add(struct m0_net_buffer *nb)
 {
-	int        result = 0;
-	/*
- 	* TODO:
- 	*   fi_send/fi_recv
- 	* */
-	return M0_RC(result);
+	struct m0_fab__buf      *fbp = nb->nb_xprt_private;
+	struct m0_fab__tm       *ma  = libfab_buf_ma(fbp);
+	struct fid_mr           *mr;
+	struct m0_fab__ep       *ep;
+	struct m0_fab__bdesc    *peer;
+	fi_addr_t                remote_rx_addr;
+	int                      qt;
+	int                      ret = 0;
+	ssize_t                  cnt = 0;
+
+        M0_PRE(libfab_tm_is_locked(ma) && libfab_tm_invariant(ma) &&
```

Comment:
Alignment.

Only check these conditions one by one.
Otherwise, if one of them fail, we don't know which one fails.